File: RotateArrayKSteps.swift
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

class Solution {
    func rotate(_ nums: inout [Int], _ k: Int) {
        let length = nums.count
        if length == 0 || k == 0 {
            return 
        }
        var steps = k%length
        reverse(&nums,0,length-1)
        reverse(&nums,0,steps-1)
        reverse(&nums,steps,length-1)
    }
    
    func reverse(_ nums: inout [Int], _ left: Int, _ right: Int) {
        var l = left
        var r = right
        while l<r {
            let temp = nums[r]
            nums[r] = nums[l]
            nums[l] = temp
            l += 1
            r -= 1
        }
    }
    
}

File: Rotate_Array_k_steps.java
// TC : O(n)
// SC : O(1)
// Method : reversal

class Solution {
    
    private int[] swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
        return nums;
    }
    public void rotate(int[] nums, int k) {
        
        int n = nums.length;
        if(n == 0 || nums == null) return;
        
        for(int i = 0; i < n/2; i++) {
            swap(nums, i, n - i - 1);
        }
        
        k = k%n;
        
        for(int i = 0; i < k/2; i++) {
                swap(nums, i, k - i - 1);
        }
        n = n + k;
        for(int i = k; i < n/2; i++) {
            swap(nums, i, n - i - 1);
        }

    }
}
